File: src/main/java/ru/hollowhorizon/mastertech/events/HollowEventHandler.java
package ru.hollowhorizon.mastertech.events;

import net.minecraft.block.Block;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.entity.player.EntityPlayerMP;
import net.minecraft.init.Blocks;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.stats.StatList;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.DimensionType;
import net.minecraft.world.World;
import net.minecraft.world.WorldServer;
import net.minecraftforge.client.event.ModelRegistryEvent;
import net.minecraftforge.event.RegistryEvent;
import net.minecraftforge.event.entity.player.ItemTooltipEvent;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.common.eventhandler.EventPriority;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import net.minecraftforge.fml.common.gameevent.PlayerEvent;
import net.minecraftforge.registries.IForgeRegistry;
import ru.hollowhorizon.mastertech.MasterTech;
import ru.hollowhorizon.mastertech.api.model.IModeled;
import ru.hollowhorizon.mastertech.registry.MTRegistry;
import ru.hollowhorizon.mastertech.util.SpawnHelper;

@Mod.EventBusSubscriber(modid = MasterTech.MODID)
public class HollowEventHandler {
    @SubscribeEvent(priority = EventPriority.HIGHEST)
    public static void itemReg(final RegistryEvent.Register<Item> reg) {
        final IForgeRegistry<Item> fr = reg.getRegistry();

        for (Item item : MTRegistry.ITEMS) {
            fr.register(item);
        }
    }

    @SubscribeEvent(priority = EventPriority.HIGHEST)
    public static void blockReg(RegistryEvent.Register<Block> reg) {
        final IForgeRegistry<Block> fr = reg.getRegistry();

        for (Block block : MTRegistry.BLOCKS.keySet()) {
            fr.register(block);
        }
    }

    @SubscribeEvent
    public static void modelRegister(ModelRegistryEvent _e) {
        for (Item item : MTRegistry.ITEMS) {
            if (item instanceof IModeled) {
                ((IModeled) item).registerModels();
            }
        }

        for (Block block : MTRegistry.BLOCKS.keySet()) {
            if (block instanceof IModeled) {
                ((IModeled) block).registerModels();
            }
        }
    }

    @SubscribeEvent
    public static void onPlayerJoining(PlayerEvent.PlayerLoggedInEvent event) {
        EntityPlayer player = event.player;

        if (!player.world.isRemote) {
            EntityPlayerMP sp = (EntityPlayerMP) player;

            int timePlayed = sp.getStatFile().readStat(StatList.PLAY_ONE_MINUTE);
            if (timePlayed == 0) {
                SpawnHelper.teleportToDim(sp, DimensionType.NETHER.getId());
                if (player.dimension == DimensionType.NETHER.getId()) {
                    WorldServer serverWorld = sp.getServerWorld();
                    buildSafeCube(serverWorld, sp.posX, sp.posY, sp.posZ);
                    BlockPos pos = pos(sp.posX, sp.posY, sp.posZ);
                    sp.setSpawnPoint(pos, true);
                    sp.setSpawnDimension(DimensionType.NETHER.getId());
                    sp.setSpawnChunk(pos, true, DimensionType.NETHER.getId());
                }
            }
        }
    }

    @SubscribeEvent
    public static void addTooltips(ItemTooltipEvent e) {
        ItemStack stack = e.getItemStack();
        if (stack.getItem() == MTRegistry.ICHOR)
            e.getToolTip().add("msg.mastertech.ichor");
        else if (stack.getItem() == MTRegistry.ICHOR_BAG)
            e.getToolTip().add("msg.mastertech.ichor_bag");
    }

    private static void nullPlace(World world, BlockPos... pos) {
        for (BlockPos pos1 : pos) {
            place(world, Blocks.AIR, pos1);
        }
    }

    private static void placeNetherrack(World world, BlockPos... pos) {
        for (BlockPos pos1 : pos) {
            place(world, Blocks.NETHERRACK, pos1);
        }
    }
    private static void place(World world, Block block, BlockPos pos) {
        world.setBlockState(pos, block.getBlockState().getBaseState(), 3);
    }

    private static void buildSafeCube(WorldServer wld, Number x, Number y, Number z) {
        int x1 = (x.intValue()) + 1;
        int x2 = (x.intValue()) + 2;
        int x_1 = (x.intValue()) - 1;
        int x_2 = (x.intValue()) - 2;

        int y1 = (y.intValue()) + 1;
        int y2 = (y.intValue()) + 2;
        int y_1 = (y.intValue()) - 1;
        int y_2 = (y.intValue()) - 2;

        int z1 = (z.intValue()) + 1;
        int z2 = (z.intValue()) + 2;
        int z_1 = (z.intValue()) - 1;
        int z_2 = (z.intValue()) - 2;
        // Netherrack wall place
        placeNetherrack(
                wld,
                //ABOVE PLATFORM
                pos(x_2, y2, z_2), pos(x_1, y2, z_2), pos(x, y2, z_2), pos(x1, y2, z_2), pos(x2, y2, z_2),
                pos(x_2, y2, z_1), pos(x_1, y2, z_1), pos(x, y2, z_1), pos(x1, y2, z_1), pos(x2, y2, z_1),
                pos(x_2, y2, z), pos(x_1, y2, z), pos(x, y2, z), pos(x1, y2, z), pos(x2, y2, z),
                pos(x_2, y2, z1), pos(x_1, y2, z1), pos(x, y2, z1), pos(x1, y2, z1), pos(x2, y2, z1),
                pos(x_2, y2, z2), pos(x_1, y2, z2), pos(x, y2, z2), pos(x1, y2, z2), pos(x2, y2, z2),
                // Y: 1 FRAME
                pos(x_2, y1, z_2), pos(x_1, y1, z_2), pos(x, y1, z_2), pos(x1, y1, z_2), pos(x2, y1, z_2),
                pos(x_2, y1, z_1), pos(x2, y1, z_1),
                pos(x_2, y1, z), pos(x2, y1, z),
                pos(x_2, y1, z1), pos(x2, y1, z1),
                pos(x_2, y1, z2), pos(x_1, y1, z2), pos(x, y1, z2), pos(x1, y1, z2), pos(x2, y1, z2),
                // Y: 0 FRAME
                pos(x_2, y, z_2), pos(x_1, y, z_2), pos(x, y, z_2), pos(x1, y, z_2), pos(x2, y, z_2),
                pos(x_2, y, z_1), pos(x2, y, z_1),
                pos(x_2, y, z), pos(x2, y, z),
                pos(x_2, y, z1), pos(x2, y, z1),
                pos(x_2, y, z2), pos(x_1, y, z2), pos(x, y, z2), pos(x1, y, z2), pos(x2, y, z2),
                // Y: -1 FRAME
                pos(x_2, y_1, z_2), pos(x_1, y_1, z_2), pos(x, y_1, z_2), pos(x1, y_1, z_2), pos(x2, y_1, z_2),
                pos(x_2, y_1, z_1), pos(x2, y_1, z_1),
                pos(x_2, y_1, z), pos(x2, y_1, z),
                pos(x_2, y_1, z1), pos(x2, y_1, z1),
                pos(x_2, y_1, z2), pos(x_1, y_1, z2), pos(x, y_1, z2), pos(x1, y_1, z2), pos(x2, y_1, z2),
                //BELOW PLATFORM
                pos(x_2, y_2, z_2), pos(x_1, y_2, z_2), pos(x, y_2, z_2), pos(x1, y_2, z_2), pos(x2, y_2, z_2),
                pos(x_2, y_2, z_1), pos(x_1, y_2, z_1), pos(x, y_2, z_1), pos(x1, y_2, z_1), pos(x2, y_2, z_1),
                pos(x_2, y_2, z), pos(x_1, y_2, z), pos(x1, y_2, z), pos(x2, y_2, z),
                pos(x_2, y_2, z1), pos(x_1, y_2, z1), pos(x, y_2, z1), pos(x1, y_2, z1), pos(x2, y_2, z1),
                pos(x_2, y_2, z2), pos(x_1, y_2, z2), pos(x, y_2, z2), pos(x1, y_2, z2), pos(x2, y_2, z2)
        );
        // Void cube place
        nullPlace(
                wld,
                // Y: 1 CUBE
                pos(x_1, y1, z_1), pos(x, y1, z_1), pos(x1, y1, z_1),
                pos(x_1, y1, z), pos(x, y1, z), pos(x1, y1, z),
                pos(x_1, y1, z1), pos(x, y1, z1), pos(x1, y1, z1),
                // Y: 0 CUBE
                pos(x_1, y, z_1), pos(x, y, z_1), pos(x1, y, z_1),
                pos(x_1, y, z), pos(x, y, z), pos(x1, y, z),
                pos(x_1, y, z1), pos(x, y, z1), pos(x1, y, z1),
                // Y: -1 CUBE
                pos(x_1, y_1, z_1), pos(x, y_1, z_1), pos(x1, y_1, z_1),
                pos(x_1, y_1, z), pos(x, y_1, z), pos(x1, y_1, z),
                pos(x_1, y_1, z1), pos(x, y_1, z1), pos(x1, y_1, z1)
        );

        //central netherbrick block, below than player
        place(wld, Blocks.NETHER_BRICK, pos(x, y_2, z));
    }

    public static BlockPos pos(Number x, Number y, Number z) {
        return new BlockPos(x.intValue(), y.intValue(), z.intValue());
    }
}

File: src/main/java/ru/hollowhorizon/mastertech/events/HollowEventHandler.kt
package ru.hollowhorizon.mastertech.events

import net.minecraft.block.Block
import net.minecraft.entity.player.EntityPlayerMP
import net.minecraft.init.Blocks
import net.minecraft.item.Item
import net.minecraft.stats.StatList
import net.minecraft.util.math.BlockPos
import net.minecraft.world.DimensionType
import net.minecraft.world.World
import net.minecraft.world.WorldServer
import net.minecraftforge.client.event.ModelRegistryEvent
import net.minecraftforge.event.RegistryEvent
import net.minecraftforge.event.entity.player.PlayerSleepInBedEvent
import net.minecraftforge.fml.common.Mod.EventBusSubscriber
import net.minecraftforge.fml.common.eventhandler.EventPriority
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
import net.minecraftforge.fml.common.gameevent.PlayerEvent.PlayerLoggedInEvent
import net.minecraftforge.fml.common.gameevent.PlayerEvent.PlayerRespawnEvent
import net.minecraftforge.fml.common.gameevent.TickEvent.PlayerTickEvent
import ru.hollowhorizon.mastertech.MasterTech
import ru.hollowhorizon.mastertech.api.model.IModeled
import ru.hollowhorizon.mastertech.registry.MTRegistry
import ru.hollowhorizon.mastertech.util.SpawnHelper

@Suppress("SENSELESS_COMPARISON")
@EventBusSubscriber(modid = MasterTech.MODID)
object HollowEventHandler {
    @JvmStatic
    @SubscribeEvent(priority = EventPriority.HIGHEST)
    fun itemReg(reg: RegistryEvent.Register<Item>) {
        val fr = reg.registry

        for (item in MTRegistry.ITEMS) {
            fr.register(item)
        }
    }

    @JvmStatic
    @SubscribeEvent(priority = EventPriority.HIGHEST)
    fun blockReg(reg: RegistryEvent.Register<Block>) {
        val fr = reg.registry

        for (block in MTRegistry.BLOCKS.keys) {
            fr.register(block)
        }
    }

    @JvmStatic
    @SubscribeEvent
    fun modelRegister(e: ModelRegistryEvent) {
        for (item in MTRegistry.ITEMS) {
            if (item is IModeled) {
                (item as IModeled).registerModels()
            }
        }

        for (block in MTRegistry.BLOCKS.keys) {
            if (block is IModeled) {
                (block as IModeled).registerModels()
            }
        }
    }

    @JvmStatic
    @SubscribeEvent
    fun onPlayerJoining(event: PlayerLoggedInEvent) {
        val player = event.player

        if (!player.world.isRemote) {
            val sp = player as EntityPlayerMP

            val timePlayed = sp.statFile.readStat(StatList.PLAY_ONE_MINUTE)
            if (timePlayed == 0) {
                SpawnHelper.teleportToDim(sp, DimensionType.NETHER.id)
            }
        }
    }

    @JvmStatic
    @SubscribeEvent
    fun onPlayerRespawn(e: PlayerRespawnEvent) {
        val player = e.player
        val level = player.world

        if (!level.isRemote) {
            if (player is EntityPlayerMP) {
                val bedLocation: BlockPos? =
                    if (player.dimension == DimensionType.OVERWORLD.id) player.getBedLocation(DimensionType.OVERWORLD.id)
                    else {
                        val dim = player.getBedLocation(player.dimension)
                        if (dim != null) dim
                        else player.getBedLocation(DimensionType.OVERWORLD.id)
                    }

                // Fuck MC 1.12.2.
                // Mojang and Lex, I hate you!
                // Fuck you.
                if (bedLocation == null) {
                    if (player.dimension != DimensionType.NETHER.id) {
                        SpawnHelper.teleportToDim(player, DimensionType.NETHER.id)
                    }
                }
            }
        }
    }

    @JvmStatic
    private fun nullPlace(world: World, vararg pos: BlockPos) {
        pos.forEach { place(world, Blocks.AIR, it) }
    }

    private fun placeNetherrack(world: World, vararg pos: BlockPos) {
        pos.forEach { place(world, Blocks.NETHERRACK, it) }
    }

    private fun place(world: World, block: Block, pos: BlockPos) {
        world.setBlockState(pos, block.blockState.baseState, 3)
    }

    private fun buildSafeCube(wld: WorldServer, x: Number, y: Number, z: Number) {
        val positions = mutableListOf<BlockPos>()
        val baseX = x.toInt()
        val baseY = y.toInt()
        val baseZ = z.toInt()

        // Collect positions for Netherrack
        for (dx in -2 .. 2) {
            for (dy in -2 .. 2) {
                for (dz in -2 .. 2) {
                    positions.add(pos(baseX + dx, baseY + dy, baseZ + dz))
                }
            }
        }
        placeNetherrack(wld, *positions.toTypedArray())

        // Collect positions for Air
        positions.clear()

        for (dy in -1 .. 1) {
            for (dx in -1 .. 1) {
                for (dz in -1 .. 1) {
                    positions.add(pos(baseX + dx, baseY + dy, baseZ + dz))
                }
            }
        }
        nullPlace(wld, *positions.toTypedArray())

        // Place Nether Brick
        place(wld, Blocks.NETHER_BRICK, pos(baseX, baseY - 2, baseZ))
    }

    private fun pos(x: Number, y: Number, z: Number): BlockPos {
        return BlockPos(x.toInt(), y.toInt(), z.toInt())
    }
}
